### DISCUSSION TITLE
Artifactory auth working when configured but not working in galaxy request

### DISCUSSION THREAD

**Discussion by onefourfive** [briantist/galactory#141] [Q&A] [answered]
Big fan of this project, very excited to get it up and running.

I stood up a POC container pointing to my Artifactory instance, but I'm experiencing a difference between configuring the API key or access token in Galactory vs in a Galaxy request. What did I get wrong?

## Info

Version: 0.11.2
Platform: docker

## What works

Configuring `artifactory-api-key` in galactory.conf

**galactory.conf**

```text
artifactory-api-key=MyArtifactoryToken...
artifactory-path=https://artifactory....
api-version=v3
upload-format=base64
```

```text
ansible-galaxy collection publish my_demo-collection_demo-1.0.0.tar.gz --token MyArtifactoryToken
Publishing collection artifact '/home/me/git/collection_demo/my_demo-collection_demo-1.0.0.tar.gz' to http://galactory.demo.org:8088/api
Collection has been published to the Galaxy server http://galactory.demo.org:8088/api
Waiting until Galaxy import task http://galactory.demo.org:8088/api/v3/imports/collections/0/ has completed
Collection has been successfully published and imported to the Galaxy server http://galactory.demo.org:8088/api
```

Collection successfully placed in Artifactory.

## What doesn't work

Using Artifactory key in Galaxy request - `401` result

**galactory.conf**

```diff
- artifactory-api-key=MyArtifactoryToken...
+ use-galaxy-auth=True
+ galaxy-auth-type=api_token
artifactory-path=https://artifactory....
api-version=v3
upload-format=base64
```

```text
ansible-galaxy collection publish my_demo-collection_demo-1.0.0.tar.gz --token MyArtifactoryToken
Publishing collection artifact '/home/me/git/collection_demo/my_demo-collection_demo-1.0.0.tar.gz' to http://galactory.demo.org:8088/api
ERROR! Error when publishing collection to (http://galactory.demo.org:8088/api) (HTTP Code: 401, Message: UNAUTHORIZED Code: Unknown)
```

No change to Artifactory.

### Debug output

With `log-headers=True`:

```text
galactory-1  | DEBUG:galactory:Headers: Accept-Encoding: identity
galactory-1  | Host: galactory.demo.org:8088
galactory-1  | User-Agent: ansible-galaxy/2.16.3 (Linux; python:3.12.3)
galactory-1  | Connection: close
galactory-1  |
galactory-1  |
galactory-1  | DEBUG:galactory:Body: b''
galactory-1  | INFO:werkzeug:10.126.12.65 - - [24/May/2024 00:58:29] "GET / HTTP/1.1" 200 -
galactory-1  | DEBUG:galactory:Headers: Accept-Encoding: identity
galactory-1  | Host: galactory.demo.org:8088
galactory-1  | User-Agent: ansible-galaxy/2.16.3 (Linux; python:3.12.3)
galactory-1  | Connection: close
galactory-1  |
galactory-1  |
galactory-1  | DEBUG:galactory:Body: b''
galactory-1  | INFO:werkzeug:10.126.12.65 - - [24/May/2024 00:58:29] "GET /api HTTP/1.1" 200 -
galactory-1  | DEBUG:galactory:Headers: Accept-Encoding: identity
galactory-1  | Host: galactory.demo.org:8088
galactory-1  | User-Agent: ansible-galaxy/2.16.3 (Linux; python:3.12.3)
galactory-1  | Content-Type: multipart/form-data; boundary="===============7353633329176979880=="
galactory-1  | Content-Length: 44480
galactory-1  | Connection: close
...
galactory-1  | DEBUG:artifactory:Raw URL passed for encoding: https://artifactory.demo.com/artifactory/galactory-generic-local/my_demo-collection_demo-1.0.0.tar.gz
galactory-1  | DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): artifactory.demo.com:443
galactory-1  | DEBUG:urllib3.connectionpool:https://artifactory.demo.com:443 "PUT /artifactory/galactory-generic-local/my_demo-collection_demo-1.0.0.tar.gz;collection_info=...;fqcn=my_demo.collection_demo;name=collection_demo;namespace=my_demo;version=1.0.0 HTTP/1.1" 401 None
galactory-1  | DEBUG:galactory:401 Client Error:  for url: https://artifactory.demo.com/artifactory/galactory-generic-local/my_demo-collection_demo-1.0.0.tar.gz;collection_info=...;fqcn=my_demo.collection_demo;name=collection_demo;namespace=my_demo;version=1.0.0
galactory-1  | INFO:werkzeug:10.126.12.65 - - [24/May/2024 00:58:29] "POST /api/v3/artifacts/collections/ HTTP/1.1" 401 -
```

**Comment by briantist** [comment]
> Big fan of this project, very excited to get it up and running.

Thanks very much! Glad to hear it. I'd love to know how you heard about it as well.

---

I can't do any deep debugging right now but I'll try to take a closer look as soon as I can.

I appreciate the detailed report! Could I confirm a few details?

```
galaxy-auth-type=api_token
```

This value should be either `api_key` or `access_token`, I'm assuming this was a copy/paste or other error in the report and not in your real configuration, because I _think_ this would have failed startup?

Also it seems like you're intending to use an API token and not an access key generally, is that correct?

Is there anything else in the galactory debug output that shows info about the auth process?

It might also be helpful to run `ansible-galaxy` with `-vvvvv` and with the environment variable `ANSIBLE_DEBUG=true` so we can take a look at the request process from the client side and maybe correlate it to something.

**Reply by onefourfive** [reply]
> Thanks very much! Glad to hear it. I'd love to know how you heard about it as well.

:) An artifactory-expert/ansible-beginner coworker let me know about it. Not sure how they found it, but this is one of the top Google results for 'artifactory galaxy'.

> This value should be either api_key or access_token, I'm assuming this was a copy/paste or other error in the report and not in your real configuration, because I think this would have failed startup?

You're right, this was a mistype on my part creating this discussion. It is `access_token`.

```diff
use-galaxy-auth=True
+ galaxy-auth-type=access_token
artifactory-path=https://artifactory....
api-version=v3
upload-format=base64
```

> Is there anything else in the galactory debug output that shows info about the auth process?

I don't believe so. Here is the full output with `log-level=DEBUG` and `log-headers=True`:

```text
galactory-1  | DEBUG:galactory:Headers: Accept-Encoding: identity
galactory-1  | Host: galaxy.demo.org:8088
galactory-1  | User-Agent: ansible-galaxy/2.16.3 (Linux; python:3.12.3)
galactory-1  | Connection: close
galactory-1  |
galactory-1  |
galactory-1  | INFO:werkzeug:10.126.12.65 - - [24/May/2024 17:35:50] "GET / HTTP/1.1" 200 -
galactory-1  | DEBUG:galactory:Headers: Accept-Encoding: identity
galactory-1  | Host: galaxy.demo.org:8088
galactory-1  | User-Agent: ansible-galaxy/2.16.3 (Linux; python:3.12.3)
galactory-1  | Connection: close
galactory-1  |
galactory-1  |
galactory-1  | INFO:werkzeug:10.126.12.65 - - [24/May/2024 17:35:50] "GET /api HTTP/1.1" 200 -
galactory-1  | DEBUG:galactory:Headers: Accept-Encoding: identity
galactory-1  | Host: galaxy.demo.org:8088
galactory-1  | User-Agent: ansible-galaxy/2.16.3 (Linux; python:3.12.3)
galactory-1  | Content-Type: multipart/form-data; boundary="===============0767931395111297708=="
galactory-1  | Content-Length: 44480
galactory-1  | Connection: close
galactory-1  |
galactory-1  |
galactory-1  | DEBUG:artifactory:Raw URL passed for encoding: https://artifactory.demo.com/artifactory/galactory-generic-local/my_demo-collection_demo-1.0.0.tar.gz
galactory-1  | DEBUG:urllib3.connectionpool:Starting new HTTPS connection (1): artifactory.demo.com:443
galactory-1  | DEBUG:urllib3.connectionpool:https://artifactory.demo.com:443 "PUT /artifactory/galactory-generic-local/my_demo-collection_demo-1.0.0.tar.gz;collection_info=%7B%22namespace%22%3A%20%22my_demo%22%2C%20%22name%22%3A%20%22collection_demo%22%2C%20%22version%22%3A%20%221.0.0%22%2C%20%22authors%22%3A%20%5B%22David%20Newswanger%22%5D%2C%20%22readme%22%3A%20%22README.md%22%2C%20%22tags%22%3A%20%5B%22demo%22%2C%20%22collection%22%5D%2C%20%22description%22%3A%20%22Demo%20showing%20how%20to%20distribute%20modules%20and%20plugins%20using%20collections%20in%20Ansible%202.8%22%2C%20%22license%22%3A%20%5B%22MIT%22%5D%2C%20%22license_file%22%3A%20null%2C%20%22dependencies%22%3A%20%7B%7D%2C%20%22repository%22%3A%20%22https%3A//github.com/newswangerd/collection_demo%22%2C%20%22documentation%22%3A%20null%2C%20%22homepage%22%3A%20null%2C%20%22issues%22%3A%20null%7D;fqcn=my_demo.collection_demo;name=collection_demo;namespace=my_demo;version=1.0.0 HTTP/1.1" 401 None
galactory-1  | DEBUG:galactory:401 Client Error:  for url: https://artifactory.demo.com/artifactory/galactory-generic-local/my_demo-collection_demo-1.0.0.tar.gz;collection_info=%7B%22namespace%22%3A%20%22my_demo%22%2C%20%22name%22%3A%20%22collection_demo%22%2C%20%22version%22%3A%20%221.0.0%22%2C%20%22authors%22%3A%20%5B%22David%20Newswanger%22%5D%2C%20%22readme%22%3A%20%22README.md%22%2C%20%22tags%22%3A%20%5B%22demo%22%2C%20%22collection%22%5D%2C%20%22description%22%3A%20%22Demo%20showing%20how%20to%20distribute%20modules%20and%20plugins%20using%20collections%20in%20Ansible%202.8%22%2C%20%22license%22%3A%20%5B%22MIT%22%5D%2C%20%22license_file%22%3A%20null%2C%20%22dependencies%22%3A%20%7B%7D%2C%20%22repository%22%3A%20%22https%3A//github.com/newswangerd/collection_demo%22%2C%20%22documentation%22%3A%20null%2C%20%22homepage%22%3A%20null%2C%20%22issues%22%3A%20null%7D;fqcn=my_demo.collection_demo;name=collection_demo;namespace=my_demo;version=1.0.0
galactory-1  | INFO:werkzeug:10.126.12.65 - - [24/May/2024 17:35:51] "POST /api/v3/artifacts/collections/ HTTP/1.1" 401
```

> It might also be helpful to run ansible-galaxy with -vvvvv and with the environment variable ANSIBLE_DEBUG=true so we can take a look at the request process from the client side and maybe correlate it to something.

Turning up verbosity for ansible-galaxy is a great idea! Unfortunately it doesn't seem to be working for me.

**ansible.cfg**

```ini
[ansible]
collections_path = .collections
debug = true
```

**ansible-galaxy version info** 

```text
$ ansible-galaxy --version
ansible-galaxy [core 2.16.3]
  config file = /home/me/git/collection_demo/ansible.cfg
  configured module search path = ['/home/me/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  ansible collection location = /home/me/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible-galaxy
  python version = 3.12.3 (main, Apr 10 2024, 05:33:47) [GCC 13.2.0] (/usr/bin/python3)
  jinja version = 3.1.2
  libyaml = True
```

**ansible-galaxy verbose output**

```text
$ ansible-galaxy -vvvvv collection publish my_demo-collection_demo-1.0.0.tar.gz  --token MyArtifactoryAccessToken
Publishing collection artifact '/home/me/git/collection_demo/my_demo-collection_demo-1.0.0.tar.gz' to ftg http://galaxy.demo.org:8088/api
ERROR! Error when publishing collection to ftg (http://galaxy.demo.org:8088/api) (HTTP Code: 401, Message: UNAUTHORIZED Code: Unknown)
```

**Reply by briantist** [reply]
For `ansible-galaxy` the position matters for verbosity. Try it like this

```
ansible-galaxy collection publish -vvvvv --no-cache my_demo-collection_demo-1.0.0.tar.gz  --token MyArtifactoryAccessToken
```
(also adding `--no-cache` to not use `ansible-galaxy`'s local response cache)

---

> You're right, this was a mistype on my part creating this discussion. It is `access_token`

In your original working config file, you're using an API key and not an access token:

```
artifactory-api-key=MyArtifactoryToken...
```

Let's confirm which you're actually using with Artifactory. The access tokens are much longer (they are JWTs) and are generated (and used) in a different way.

So if what you have is actually an API key, then the original config working makes sense, and the new config not working also makes sense, since you're telling galactory that the auth included in the galaxy request is an access token (which it will then treat that way).

If all the above is correct, and the configured auth was indeed using an API token successfully, then changing the request-based config to use `galaxy-auth-type=api_key` should work.

If the configured auth was using an API key successfully, but your intent was to change to an access token at the same time as changing to request-provided auth, I would recommend doing one change at a time. Try request-based API key (using the working key you had before), and/or try configured auth with the same access token you're trying now.

**Reply by onefourfive** [reply]
Great, thanks for the detailed response!

ansible-galaxy likes `-vvvvv` elsewhere, got it. It's not familiar with `--no-cache` though.

```text
$ ansible-galaxy collection publish -vvvvv --no-cache my_demo-collection_demo-1.0.0.tar.gz --token MyArtifactoryToken
usage: ansible-galaxy [-h] [--version] [-v] TYPE ...
ansible-galaxy: error: unrecognized arguments: --no-cache
```

but here's the proper verbose output:

```text
$ ansible-galaxy collection publish -vvvvv my_demo-collection_demo-1.0.0.tar.gz --token MyArtifactoryToken
ansible-galaxy [core 2.16.3]
  config file = /home/me/git/collection_demo/ansible.cfg
  configured module search path = ['/home/me/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  ansible collection location = /home/me/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible-galaxy
  python version = 3.12.3 (main, Apr 10 2024, 05:33:47) [GCC 13.2.0] (/usr/bin/python3)
  jinja version = 3.1.2
  libyaml = True
Using /home/me/git/collection_demo/ansible.cfg as config file
Initial connection to galaxy_server: http://galaxy.demo.org:8088
Opened /home/me/.ansible/galaxy_token
Calling Galaxy at http://galaxy.demo.org:8088
Calling Galaxy at http://galaxy.demo.org:8088/api
Found API version 'v3' with Galaxy server (http://galaxy.demo.org:8088/api)
Publishing collection artifact '/home/me/git/collection_demo/my_demo-collection_demo-1.0.0.tar.gz' to http://galaxy.demo.org:8088/api
Calling Galaxy at http://galaxy.demo.org:8088/api/v3/artifacts/collections/
ERROR! Error when publishing collection to (http://galaxy.demo.org:8088/api) (HTTP Code: 401, Message: UNAUTHORIZED Code: Unknown)
```

Regarding tokens: I'm using an Artifactory Identity Token. I generated it at My Profile > Identity Tokens, and its 65 characters long

```text
echo $ARTIFACTORY_TOKEN | wc -m
65
```

I gave it a shot with the API key though, but it doesn't seem to change the outcome.

```diff
+galaxy-auth-type=api_key
```

```text
echo $ARTIFACTORY_APIKEY | wc -m
74
```

```text
$ ansible-galaxy collection publish -vvvvv my_demo-collection_demo-1.0.0.tar.gz -s ftg --token $ARTIFACTORY_APIKEY
ansible-galaxy [core 2.16.3]
  config file = /home/me/git/collection_demo/ansible.cfg
  configured module search path = ['/home/me/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  ansible collection location = /home/me/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible-galaxy
  python version = 3.12.3 (main, Apr 10 2024, 05:33:47) [GCC 13.2.0] (/usr/bin/python3)
  jinja version = 3.1.2
  libyaml = True
Using /home/me/git/collection_demo/ansible.cfg as config file
Initial connection to galaxy_server: http://galaxy.demo.org:8088
Opened /home/me/.ansible/galaxy_token
Calling Galaxy at http://galaxy.demo.org:8088
Calling Galaxy at http://galaxy.demo.org:8088/api
Found API version 'v3' with Galaxy server ftg (http://galaxy.demo.org:8088/api)
Publishing collection artifact '/home/me/git/collection_demo/my_demo-collection_demo-1.0.0.tar.gz' to ftg http://galaxy.demo.org:8088/api
Calling Galaxy at http://galaxy.demo.org:8088/api/v3/artifacts/collections/
ERROR! Error when publishing collection to ftg (http://galaxy.demo.org:8088/api) (HTTP Code: 401, Message: UNAUTHORIZED Code: Unknown)
```

**Reply by briantist** [reply]
> It's not familiar with `--no-cache` though.

Oh right, I think it doesn't  use the response cache with the `publish` subcommand.

> I gave it a shot with the API key though, but it doesn't seem to change the outcome.

I think at this point there's a few ways we could go. One is to try making direct requests to Artifactory with `curl` or something similar, setting the auth as needed in a header. I don't have the exact command handy though.

Another is to make a small python script to test out uploading to Artifactory using the [same library galactory uses](https://github.com/devopshq/artifactory).

Another option is to start adding more logging to galactory so that we can figure out where the problem might be, but this would take a bit longer, especially for me since I have very little time lately. If you're up for it you could make some logging changes locally. It's pretty easy to run galactory from a venv, and even to step through the code as it's running. I cannot reproduce the issue myself though so while I could in theory implement logging, I can't do in-depth debugging for this particular issue.

**Reply by onefourfive** [reply]
Thanks for the tip with the library-- I started to get cracking on it and I ran into an issue it seems you yourself have reported! This might be a bit circular but were you able to work around this? devopshq/artifactory#430

**Reply by briantist** [reply]
ah, there is no way to work around that until someone does the mentioned refactor in that library. You'll need to use a supported Python version. Here are the ones I test Artifactory against (trying to add 3.12 is how I caught the above issue):
- https://github.com/briantist/galactory/blob/main/.github/workflows/test.yml#L20-L23

**Reply by onefourfive** [reply]
Ok, I've taken some more time to debug this and I think I'm at a interesting point: there is no 'Authorization' header being sent by the `ansible-galaxy` cli. This is in fact, the complete uncensored header:

```text
galactory-1  | DEBUG:galactory:Headers: Accept-Encoding: identity
galactory-1  | Host: galactory.demo.org:8088
galactory-1  | User-Agent: ansible-galaxy/2.16.3 (Linux; python:3.12.3)
galactory-1  | Content-Type: multipart/form-data; boundary="===============7353633329176979880=="
galactory-1  | Content-Length: 44480
galactory-1  | Connection: close
```

I inserted some extra lines into the `authorize()` function in utilities.py to print the Authorization header which came out as `None` (but the `log-headers` option was already showing me this!)

The next step for me will be to understand what [ansible-galaxy](https://github.com/ansible/ansible/tree/v2.16.3/lib/ansible/galaxy) is doing with headers / the `--token` option, but perhaps you're better versed in that code base? Either way I'm diving in.

**Reply by briantist** [reply]
ahhh I just had an idea: maybe (similar to the verbosity option) it's a matter of position? In your OP you have `--token` after the collection path, but it should probably be before

```
$ ansible-galaxy collection publish --help
usage: ansible-galaxy collection publish [-h] [-s API_SERVER] [--token API_KEY] [-c] [-v] [--no-wait]
                                         [--import-timeout IMPORT_TIMEOUT]
                                         collection_path
```

**Reply by onefourfive** [reply]
oh would that it were so simple! I gave that a try, order matching the ansible-galaxy help usage:

```text
ansible-galaxy collection publish -s local --token $ARTAPI_TOKEN my_demo-collection_demo-1.0.0.tar.gz
```

but that yields the same `Authorization`-less header it seems

```text
galactory-1  | INFO:werkzeug:10.10.12.65 - - [31/May/2024 22:15:51] "GET /api HTTP/1.1" 200 -
galactory-1  | INFO:galactory:authorize() called.
galactory-1  |   apikey=  None
galactory-1  |   auth header= None
galactory-1  | INFO:galactory:header info: Accept-Encoding: identity
galactory-1  | Host: galaxy.demo.org:8088
galactory-1  | User-Agent: ansible-galaxy/2.16.3 (Linux; python:3.12.3)
galactory-1  | Content-Type: multipart/form-data; boundary="===============9031658227851557661=="
galactory-1  | Content-Length: 44480
galactory-1  | Connection: close
```

**Comment by onefourfive** [comment, marked as answer]
Answering my own question here-- this an open bug in ansible since (at least) 2.15! ansible/ansible#81698 

For a token to even be included in the request to a private / custom Galaxy server, it must be specified in ansible.cfg explicitly.

## Working:

```ini
[galaxy]
server_list = local

[galaxy_server.local]
url = "http://galaxy.demo.org:8088/"
token = WholeTokenExplicitlyHere
```

```text
ansible-galaxy collection publish collection-ver.tgz
```

## Not working:

* Using `token_path` in `ansible.cfg` in any case, even with `--token` on commandline

In this case no token is provided and the `Authorization` header is never sent.

**Reply by briantist** [reply]
Nice catch! It was also suggested in that issue that you can use the environment variable.

Another option that's even less documented is to put the token in a file in `~/.ansible/galaxy_token`.

Despite the lack of extension, that file is actually YAML, and its format is like this:

```yaml
token: WholeTokenExplicitlyHere
```

This is actually the file I used for my local testing way back, but I did not expect `--token` to silently be ignored.

It seems that this file might actually not be recommended though according to this:
- https://github.com/ansible/ansible/issues/82260#issuecomment-1823001779

The way I found out about it originally was with verbose output on `ansible-galaxy`

```
Initial connection to galaxy_server: https://galaxy.ansible.com
Opened /home/briantist/.ansible/galaxy_token
Calling Galaxy at https://galaxy.ansible.com/api/
```

---

Anyway, I'm glad this was figured out, and I'm sorry for the trouble. I hope the experience with galactory is smoother from here on out!
